src/Razor/src/Microsoft.VisualStudio.Editor.Razor/TelemetryReporter.cs

```diff
@@ -207,4 +211,47 @@ private static TelemetrySeverity ToTelemetrySeverity(Severity severity)
             Severity.High => TelemetrySeverity.High,
             _ => throw new InvalidOperationException($"Unknown severity: {severity}")
         };
+
+    public IDisposable ReportScopedEvent(string name, Severity severity)
+    {
+        return new TelemetryScope(this, name, severity, ImmutableDictionary<string, object?>.Empty);
+    }
+
+    public IDisposable ReportScopedEvent<T>(string name, Severity severity, ImmutableDictionary<string, T> values)
+    {
+        return new TelemetryScope(this, name, severity, values.ToImmutableDictionary((tuple) => tuple.Key, (tuple) => (object?)tuple.Value));
+    }
+
+    private class TelemetryScope : IDisposable
+    {
+        private readonly ITelemetryReporter _telemetryReporter;
+        private string _name;
+        private Severity _severity;
+        private ImmutableDictionary<string, object?> _values;
+        private bool _disposed;
+        private Stopwatch _stopwatch;
+
+        public TelemetryScope(ITelemetryReporter telemetryReporter, string name, Severity severity, ImmutableDictionary<string, object?> values)
+        {
+            _telemetryReporter = telemetryReporter;
+            _name = name;
+            _severity = severity;
+            _values = values;
+            _stopwatch = Stopwatch.StartNew();
```

Comment:
There is an existing object pool of stop watches. Since this is disposable, consider using a pooled stopwatch rather than allocating another one. Then, return it to the pool in Dispose below.